linkerd/service-profiles/src/http/proxy.rs

```diff
@@ -0,0 +1,136 @@
+
+/// A router that uses a per-route `Proxy` to wrap a common underlying
+/// `Service`.
+///
+/// This router is similar to `linkerd_stack::NewRouter` and
+/// `linkerd_cache::Cache` with a few differences:
+///
+/// * It's `Proxy`-specific;
+/// * Routes are constructed eagerly as the profile updates;
+/// * Routes are removed eagerly as the profile updates (i.e. there's no
+///   idle-oriented eviction).
+#[derive(Clone, Debug)]
+pub struct NewProxyRouter<M, N> {
+    new_proxy: M,
+    new_service: N,
+}
+
+#[derive(Debug)]
+pub struct ProxyRouter<T, N, P, S> {
+    new_proxy: N,
+    inner: S,
+    target: T,
+    rx: ReceiverStream,
+    http_routes: Vec<(RequestMatch, Route)>,
+    proxies: HashMap<Route, P>,
+}
+
+// === impl NewProxyRouter ===
+
+impl<M: Clone, N> NewProxyRouter<M, N> {
+    pub fn layer(new_proxy: M) -> impl layer::Layer<N, Service = Self> + Clone {
+        layer::mk(move |new_service| Self {
+            new_service,
+            new_proxy: new_proxy.clone(),
+        })
+    }
+}
+
+impl<T, M, N> NewService<T> for NewProxyRouter<M, N>
+where
+    T: Param<Receiver> + Clone,
+    N: NewService<T> + Clone,
+    M: NewService<(Route, T)> + Clone,
+{
+    type Service = ProxyRouter<T, M, M::Service, N::Service>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        let rx = target.param();
+        let inner = self.new_service.new_service(target.clone());
+        ProxyRouter {
+            inner,
+            target,
+            rx: rx.into(),
+            http_routes: Vec::new(),
+            proxies: HashMap::new(),
+            new_proxy: self.new_proxy.clone(),
+        }
+    }
+}
+
+// === impl ProxyRouter ===
+
+type ProxyResponseFuture<F1, E1, F2, E2> =
+    future::Either<future::MapErr<F1, fn(E1) -> Error>, future::MapErr<F2, fn(E2) -> Error>>;
+
+impl<B, T, N, P, S, Rsp> Service<http::Request<B>> for ProxyRouter<T, N, P, S>
+where
+    T: Clone,
+    N: NewService<(Route, T), Service = P> + Clone,
+    P: Proxy<http::Request<B>, S, Request = http::Request<B>, Response = Rsp>,
+    S: Service<http::Request<B>, Response = Rsp>,
+    S::Error: Into<Error>,
+{
+    type Response = Rsp;
+    type Error = Error;
+    type Future = ProxyResponseFuture<S::Future, S::Error, P::Future, P::Error>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<()>> {
+        // Poll the inner service first so we don't bother updating routes unless we can actually
+        // use them.
+        futures::ready!(self.inner.poll_ready(cx).map_err(Into::into))?;
+
+        // If the routes have been updated, update the cache.
+        if let Poll::Ready(Some(Profile { http_routes, .. })) = self.rx.poll_next_unpin(cx) {
+            debug!(routes = %http_routes.len(), "Updating HTTP routes");
+            let routes = http_routes
+                .iter()
+                .map(|(_, r)| r.clone())
+                .collect::<HashSet<_>>();
+            self.http_routes = http_routes;
```

Comment:
nit, take it or leave it: this seems like it could be
```suggestion
            self.http_routes.clear();
            self.http_routes.extend(routes.iter().map(|(_, r)| r.clone()));
```
which has the nice side benefit of letting us reuse the prior allocation for the hashmap if the new `routes` is the same size or smaller...not a blocker though.

Reply:
I don't follow?

`self.http_routes = http_routes;` shouldn't allocate.

The issue is that we temporarily allocate a `HashSet` so that we can do quick lookups in the following `self.proxies.retain(...)`. I think we basically need to have some sort of temporary allocation, since we can't iterate over `self.proxies` while modifying it.

Reply:
hahahaha, i can't read --- i mixed up `http_routes` and `routes`. my bad.